File: internal/tui/views/activity.go
package views

import (
	"context"
	"fmt"
	"strings"
	"time"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
	"github.com/mailersend/mailersend-cli/internal/sdkclient"
	"github.com/mailersend/mailersend-cli/internal/tui/components"
	"github.com/mailersend/mailersend-cli/internal/tui/theme"
	"github.com/mailersend/mailersend-cli/internal/tui/types"
	"github.com/mailersend/mailersend-go"
)

var (
	domainTabStyle = lipgloss.NewStyle().
			Padding(0, 2)

	activeDomainTabStyle = lipgloss.NewStyle().
				Bold(true).
				Foreground(theme.Primary).
				Background(theme.BgSelected).
				Padding(0, 2)

	domainTabBarStyle = lipgloss.NewStyle().
				Border(lipgloss.NormalBorder(), false, false, true, false).
				BorderForeground(theme.Muted).
				MarginBottom(1)
)

// ActivityView displays the activity log.
type ActivityView struct {
	client          *mailersend.Mailersend
	transport       *sdkclient.CLITransport
	table           components.Table
	detail          components.DetailPanel
	items           []types.ActivityItem
	domains         []mailersend.Domain
	activeDomainIdx int
	loading         bool
	loadingDomains  bool
	err             error
	width           int
	height          int
	focused         bool
	showingDetail   bool
}

// NewActivityView creates a new activity view.
func NewActivityView(client *mailersend.Mailersend, transport *sdkclient.CLITransport) ActivityView {
	columns := []components.Column{
		{Title: "TIME", Width: 19},
		{Title: "EVENT", Width: 14},
		{Title: "RECIPIENT", Width: 28},
		{Title: "SUBJECT", Width: 30},
	}
	table := components.NewTable(columns)
	table.SetEmptyMessage("Select a domain to view activity.")

	return ActivityView{
		client:         client,
		transport:      transport,
		table:          table,
		loading:        true,
		loadingDomains: true,
	}
}

// SetClient updates the SDK client.
func (v *ActivityView) SetClient(client *mailersend.Mailersend, transport *sdkclient.CLITransport) {
	v.client = client
	v.transport = transport
}

// SetSize sets the view dimensions.
func (v *ActivityView) SetSize(width, height int) {
	v.width = width
	v.height = height
	// Reserve space for domain bar
	v.table.SetSize(width, height-4)
}

// SetFocused sets whether this view is focused.
func (v *ActivityView) SetFocused(focused bool) {
	v.focused = focused
	v.table.SetFocused(focused)
}

// Loading returns whether the view is loading.
func (v ActivityView) Loading() bool {
	return v.loading
}

// Error returns any error.
func (v ActivityView) Error() error {
	return v.err
}

// ItemCount returns the number of items.
func (v ActivityView) ItemCount() int {
	return len(v.items)
}

// Fetch returns a command to fetch domains first, then activity.
func (v ActivityView) Fetch() tea.Cmd {
	return v.fetchDomains()
}

func (v ActivityView) fetchDomains() tea.Cmd {
	return func() tea.Msg {
		if v.client == nil {
			return types.ActivityLoadedMsg{Err: nil}
		}

		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		domains, err := sdkclient.FetchAll(ctx, func(ctx context.Context, page, perPage int) ([]mailersend.Domain, bool, error) {
			root, _, err := v.client.Domain.List(ctx, &mailersend.ListDomainOptions{
				Page:  page,
				Limit: perPage,
			})
			if err != nil {
				return nil, false, sdkclient.WrapError(v.transport, err)
			}
			return root.Data, root.Links.Next != "", nil
		}, 100)

		if err != nil {
			return ActivityDomainsLoadedMsg{Err: err}
		}

		return ActivityDomainsLoadedMsg{Domains: domains}
	}
}

// ActivityDomainsLoadedMsg is sent when domains are loaded for the activity view.
type ActivityDomainsLoadedMsg struct {
	Domains []mailersend.Domain
	Err     error
}

func (v ActivityView) fetchActivity() tea.Cmd {
	return func() tea.Msg {
		if v.client == nil || len(v.domains) == 0 {
			return types.ActivityLoadedMsg{Err: nil}
		}

		domainID := v.domains[v.activeDomainIdx].ID

		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		// Use last 30 days
		now := time.Now()
		dateFrom := now.AddDate(0, 0, -30).Unix()
		dateTo := now.Unix()

		items, err := sdkclient.FetchAll(ctx, func(ctx context.Context, page, perPage int) ([]types.ActivityItem, bool, error) {
			root, _, err := v.client.Activity.List(ctx, &mailersend.ActivityOptions{
				DomainID: domainID,
				Page:     page,
				Limit:    perPage,
				DateFrom: dateFrom,
				DateTo:   dateTo,
			})
			if err != nil {
				return nil, false, sdkclient.WrapError(v.transport, err)
			}

			var items []types.ActivityItem
			for _, d := range root.Data {
				item := types.ActivityItem{
					ID:        d.ID,
					CreatedAt: d.CreatedAt,
					Type:      d.Type,
				}
				item.Email.From = d.Email.From
				item.Email.To = d.Email.Recipient.Email
				item.Email.Subject = d.Email.Subject
				items = append(items, item)
			}

			return items, root.Links.Next != "", nil
		}, 100)

		return types.ActivityLoadedMsg{
			Items: items,
			Err:   err,
		}
	}
}

// Update handles messages for this view.
func (v ActivityView) Update(msg tea.Msg) (ActivityView, tea.Cmd) {
	switch msg := msg.(type) {
	case ActivityDomainsLoadedMsg:
		v.loadingDomains = false
		if msg.Err != nil {
			v.loading = false
			v.err = msg.Err
			return v, nil
		}
		v.domains = msg.Domains
		if len(v.domains) > 0 {
			// Fetch activity for first domain
			return v, v.fetchActivity()
		}
		v.loading = false
		v.table.SetEmptyMessage("No domains found. Add a domain first.")
	case types.ActivityLoadedMsg:
		v.loading = false
		v.err = msg.Err
		if msg.Err == nil {
			v.items = msg.Items
			v.updateTable()
		}
	}
	return v, nil
}

// HandleKey handles key events when this view is active.
func (v *ActivityView) HandleKey(msg tea.KeyMsg) tea.Cmd {
	// Handle detail view navigation
	if v.showingDetail {
		switch msg.String() {
		case "esc", "backspace", "q":
			v.showingDetail = false
		}
		return nil
	}

	switch msg.String() {
	case "j", "down":
		v.table.MoveDown()
	case "k", "up":
		v.table.MoveUp()
	case "g":
		v.table.GotoTop()
	case "G":
		v.table.GotoBottom()
	case "h", "left":
		if len(v.domains) > 0 {
			v.activeDomainIdx--
			if v.activeDomainIdx < 0 {
				v.activeDomainIdx = len(v.domains) - 1
			}
			v.loading = true
			v.table.SetLoading(true)
			return v.fetchActivity()
		}
	case "l", "right":
		if len(v.domains) > 0 {
			v.activeDomainIdx = (v.activeDomainIdx + 1) % len(v.domains)
			v.loading = true
			v.table.SetLoading(true)
			return v.fetchActivity()
		}
	case "enter":
		v.showDetail()
	case "r":
		v.loading = true
		v.table.SetLoading(true)
		return v.fetchActivity()
	}
	return nil
}

// SelectedItem returns the currently selected activity item.
func (v ActivityView) SelectedItem() *types.ActivityItem {
	idx := v.table.Cursor()
	if idx >= 0 && idx < len(v.items) {
		return &v.items[idx]
	}
	return nil
}

func (v *ActivityView) showDetail() {
	item := v.SelectedItem()
	if item == nil {
		return
	}

	v.detail.SetTitle("Activity Event")

	created := item.CreatedAt
	if t, err := time.Parse(time.RFC3339, item.CreatedAt); err == nil {
		created = t.Format("2006-01-02 15:04:05")
	}

	v.detail.SetRows([]components.DetailRow{
		{Label: "ID", Value: item.ID},
		{Label: "Event Type", Value: item.Type},
		{Label: "Time", Value: created},
		{Label: "From", Value: item.Email.From},
		{Label: "To", Value: item.Email.To},
		{Label: "Subject", Value: item.Email.Subject},
	})
	v.detail.SetSize(v.width, v.height)
	v.showingDetail = true
}

func (v *ActivityView) updateTable() {
	var rows [][]string
	for _, item := range v.items {
		created := item.CreatedAt
		if t, err := time.Parse(time.RFC3339, item.CreatedAt); err == nil {
			created = t.Format("2006-01-02 15:04:05")
		}

		subject := item.Email.Subject
		if len(subject) > 30 {
			subject = subject[:27] + "..."
		}

		rows = append(rows, []string{
			created,
			item.Type,
			item.Email.To,
			subject,
		})
	}
	v.table.SetRows(rows)
	v.table.SetLoading(false)
}

// View renders the activity view.
func (v ActivityView) View() string {
	if v.showingDetail {
		return v.detail.View()
	}

	var b strings.Builder

	// Domain selector bar
	if len(v.domains) > 0 {
		var domainTabs []string
		for i, d := range v.domains {
			name := d.Name
			if len(name) > 20 {
				name = name[:17] + "..."
			}
			style := domainTabStyle
			if i == v.activeDomainIdx {
				style = activeDomainTabStyle
			}
			domainTabs = append(domainTabs, style.Render(name))
		}

		tabBar := lipgloss.JoinHorizontal(lipgloss.Top, domainTabs...)
		b.WriteString(domainTabBarStyle.Render(tabBar))
		b.WriteString("\n")

		// Hint
		hint := fmt.Sprintf("← → to switch domains | %d events (last 30 days)", len(v.items))
		b.WriteString(lipgloss.NewStyle().Foreground(theme.Muted).Render(hint))
		b.WriteString("\n\n")
	} else if v.loadingDomains {
		b.WriteString(lipgloss.NewStyle().Foreground(theme.Muted).Render("Loading domains..."))
		b.WriteString("\n\n")
	}

	// Table
	b.WriteString(v.table.View())

	return b.String()
}

// ShowingDetail returns whether the detail view is active.
func (v ActivityView) ShowingDetail() bool {
	return v.showingDetail
}

File: src/tui/views/activity.rs
use anyhow::Result;
use chrono::{Duration, Utc};
use crossterm::event::KeyCode;
use ratatui::layout::{Constraint, Layout, Rect};
use ratatui::style::{Modifier, Style};
use ratatui::text::{Line, Span};
use ratatui::widgets::Paragraph;
use ratatui::Frame;
use serde_json::Value;

use super::{fmt_datetime, KeyResult};
use crate::api::{self, Client};
use crate::tui::components::detail::{DetailPanel, DetailRow};
use crate::tui::components::table::{Cell, Column, Table};
use crate::tui::theme;
use crate::util::{jpath, jstr};

pub const EMPTY: &str = "Select a domain to view activity.";
pub const EMPTY_NO_DOMAINS: &str = "No domains found. Add a domain first.";

pub fn columns() -> Vec<Column> {
    vec![
        Column {
            title: "TIME",
            width: 19,
        },
        Column {
            title: "EVENT",
            width: 14,
        },
        Column {
            title: "RECIPIENT",
            width: 28,
        },
        Column {
            title: "SUBJECT",
            width: 30,
        },
    ]
}

pub fn fetch_domains(client: &Client) -> Result<Vec<Value>> {
    api::fetch_all_paged(client, "/domains", &[], 100)
}

pub fn fetch_activity(client: &Client, domain_id: &str) -> Result<Vec<Value>> {
    let now = Utc::now();
    let date_from = (now - Duration::days(30)).timestamp();
    let date_to = now.timestamp();
    api::fetch_all_paged(
        client,
        &format!("/activity/{domain_id}"),
        &[
            ("date_from", date_from.to_string()),
            ("date_to", date_to.to_string()),
        ],
        100,
    )
}

pub struct View {
    pub table: Table,
    detail: DetailPanel,
    pub items: Vec<Value>,
    pub domains: Vec<Value>,
    pub active_domain_idx: usize,
    pub loading: bool,
    pub loading_domains: bool,
    showing_detail: bool,
}

impl View {
    pub fn new() -> Self {
        Self {
            table: Table::new(columns(), EMPTY),
            detail: DetailPanel::new(),
            items: Vec::new(),
            domains: Vec::new(),
            active_domain_idx: 0,
            loading: true,
            loading_domains: true,
            showing_detail: false,
        }
    }

    pub fn set_domains(&mut self, domains: Vec<Value>) {
        self.loading_domains = false;
        self.domains = domains;
        self.active_domain_idx = 0;
    }

    pub fn no_domains(&mut self) {
        self.loading = false;
        self.loading_domains = false;
        self.table.set_empty_msg(EMPTY_NO_DOMAINS);
    }

    pub fn domain_id(&self) -> Option<String> {
        self.domains
            .get(self.active_domain_idx)
            .map(|d| jstr(d, "id"))
    }

    pub fn set_loaded(&mut self, items: Vec<Value>) {
        let rows = items
            .iter()
            .map(|item| {
                vec![
                    Cell::plain(fmt_datetime(&jstr(item, "created_at"))),
                    Cell::plain(jstr(item, "type")),
                    Cell::plain(jpath(item, "email.recipient.email")),
                    Cell::plain(jpath(item, "email.subject")),
                ]
            })
            .collect();
        self.items = items;
        self.table.set_rows(rows);
        self.table.set_loading(false);
        self.loading = false;
    }

    pub fn handle_key(&mut self, code: KeyCode) -> KeyResult {
        if self.showing_detail {
            if matches!(code, KeyCode::Esc | KeyCode::Backspace | KeyCode::Char('q')) {
                self.showing_detail = false;
            }
            return KeyResult::None;
        }

        match code {
            KeyCode::Char('j') | KeyCode::Down => self.table.move_down(),
            KeyCode::Char('k') | KeyCode::Up => self.table.move_up(),
            KeyCode::Char('g') => self.table.goto_top(),
            KeyCode::Char('G') => self.table.goto_bottom(),
            KeyCode::Char('h') | KeyCode::Left => {
                if !self.domains.is_empty() {
                    self.active_domain_idx = if self.active_domain_idx == 0 {
                        self.domains.len() - 1
                    } else {
                        self.active_domain_idx - 1
                    };
                    self.loading = true;
                    self.table.set_loading(true);
                    return KeyResult::Fetch;
                }
            }
            KeyCode::Char('l') | KeyCode::Right => {
                if !self.domains.is_empty() {
                    self.active_domain_idx = (self.active_domain_idx + 1) % self.domains.len();
                    self.loading = true;
                    self.table.set_loading(true);
                    return KeyResult::Fetch;
                }
            }
            KeyCode::Enter => self.show_detail(),
            KeyCode::Char('r') => {
                self.loading = true;
                self.table.set_loading(true);
                return KeyResult::Fetch;
            }
            _ => {}
        }
        KeyResult::None
    }

    fn show_detail(&mut self) {
        let Some(item) = self.items.get(self.table.cursor()) else {
            return;
        };
        self.detail.set_title("Activity Event");
        self.detail.set_rows(vec![
            DetailRow::new("ID", jstr(item, "id")),
            DetailRow::new("Event Type", jstr(item, "type")),
            DetailRow::new("Time", fmt_datetime(&jstr(item, "created_at"))),
            DetailRow::new("From", jpath(item, "email.from")),
            DetailRow::new("To", jpath(item, "email.recipient.email")),
            DetailRow::new("Subject", jpath(item, "email.subject")),
        ]);
        self.showing_detail = true;
    }

    pub fn render(&mut self, frame: &mut Frame, area: Rect, focused: bool) {
        if self.showing_detail {
            frame.render_widget(Paragraph::new(self.detail.lines()), area);
            return;
        }

        let chrome_height = if self.domains.is_empty() && !self.loading_domains {
            0
        } else {
            4
        };

        let [chrome, table_area] =
            Layout::vertical([Constraint::Length(chrome_height), Constraint::Min(0)]).areas(area);

        if !self.domains.is_empty() {
            frame.render_widget(Paragraph::new(self.domain_chrome()), chrome);
        } else if self.loading_domains {
            frame.render_widget(
                Paragraph::new(Line::styled(
                    "Loading domains...",
                    Style::new().fg(theme::MUTED),
                )),
                chrome,
            );
        }

        self.table.set_focused(focused);
        self.table
            .set_size(table_area.width as usize, table_area.height as usize);
        frame.render_widget(Paragraph::new(self.table.lines()), table_area);
    }

    fn domain_chrome(&self) -> Vec<Line<'static>> {
        let mut tabs: Vec<Span<'static>> = Vec::new();
        let mut bar_width = 0usize;
        for (i, d) in self.domains.iter().enumerate() {
            let mut name = jstr(d, "name");
            if name.chars().count() > 20 {
                name = format!("{}...", name.chars().take(17).collect::<String>());
            }
            let label = format!("  {name}  ");
            bar_width += label.chars().count();
            let style = if i == self.active_domain_idx {
                Style::new()
                    .fg(theme::PRIMARY)
                    .bg(theme::BG_SELECTED)
                    .add_modifier(Modifier::BOLD)
            } else {
                Style::new()
            };
            tabs.push(Span::styled(label, style));
        }

        vec![
            Line::from(tabs),
            Line::styled("─".repeat(bar_width), Style::new().fg(theme::MUTED)),
            Line::styled(
                format!(
                    "← → to switch domains | {} events (last 30 days)",
                    self.items.len()
                ),
                Style::new().fg(theme::MUTED),
            ),
            Line::raw(""),
        ]
    }
}
